app/controllers/Application.scala

```diff
@@ -1,71 +1,82 @@
+      new MacBodyParser(hmacHeader, secret, algorithm, maxBodySize)
   }
 
   private[controllers] class MacBodyParser(
     hmacHeader: String,
     secret: SecretKeySpec,
-    algorithm: String) extends BodyParser[Unit] {
+    algorithm: String,
+    maxBodySize: Int) extends BodyParser[Array[Byte]] {
 
     def hex2bytes(hex: String): Array[Byte] =
       hex.replaceAll("[^0-9A-Fa-f]", "").sliding(2, 2).toArray.map(Integer.parseInt(_, 16).toByte)
 
-    override def apply(request: RequestHeader): Iteratee[Array[Byte], Either[Result, Unit]] = {
+    override def apply(request: RequestHeader): Iteratee[Array[Byte], Either[Result, Array[Byte]]] = {
       val hexSignature = request.headers.get(hmacHeader).map(_.dropWhile(_ != '=').drop(1)).getOrElse("")
       val signature = hex2bytes(hexSignature)
-      Iteratee.fold[Array[Byte], Mac] {
+      Iteratee.fold[Array[Byte], (Mac, ArrayBuffer[Byte])] {
         val mac = Mac.getInstance(algorithm)
         mac.init(secret)
-        mac
-      } { (mac, bytes) =>
-        mac.update(bytes)
-        mac
+        (mac, ArrayBuffer.empty)
+      } {
+        case ((mac, buffer), bytes) =>
+          mac.update(bytes)
+          val newBuffer = if (buffer.length + bytes.length <= maxBodySize) buffer ++ bytes else buffer
+          (mac, newBuffer)
       }.map {
-        case _   if signature.isEmpty                     => Left(Results.BadRequest(s"No $hmacHeader header present"))
-        case mac if mac.doFinal().sameElements(signature) => Right(())
-        case _                                            => Left(Results.Unauthorized("Bad signature"))
+        case _ if signature.isEmpty                                 => Left(Results.BadRequest(s"No $hmacHeader header present"))
+        case (mac, buffer) if mac.doFinal().sameElements(signature) => Right(buffer.toArray)
+        case _                                                      => Left(Results.Unauthorized("Bad signature"))
       }
     }
   }
 
   private def getDocRenderer(
     host: String,
-    docRenderers: Map[String, ActorRef],
+    pathVersion: String,
+    docRenderers: Map[String, Map[String, ActorRef]],
     hostPrefixAliases: Map[String, String]): Option[ActorRef] = {
+
     val hostPrefix = host.takeWhile(c => c != '.' && c != ':')
-    docRenderers.get(hostPrefix).orElse {
-      hostPrefixAliases.get(hostPrefix) match {
-        case Some(aliasedHostPrefix) => docRenderers.get(aliasedHostPrefix)
-        case None                    => None
+    docRenderers
+      .get(hostPrefix)
+      .orElse {
+        hostPrefixAliases.get(hostPrefix) match {
+          case Some(aliasedHostPrefix) => docRenderers.get(aliasedHostPrefix)
+          case None                    => None
+        }
       }
-    }
+      .flatMap(_.get(pathVersion))
   }
 }
 
 class Application @Inject() (
   @Named("ConductRDocRenderer") conductrDocRenderer: ActorRef,
```

Comment:
Good question. We no longer need to do this. The mapping of versions to renderers should be localised to the Application. I will change this. 

> On 18 Sep 2015, at 7:06 am, Markus Jura <notifications@github.com> wrote:
> 
> In app/controllers/Application.scala:
> 
> >        }
> > -    }
> > +      .flatMap(_.get(pathVersion))
> >    }
> >  }
> >  
> >  class Application @Inject() (
> >    @Named("ConductRDocRenderer") conductrDocRenderer: ActorRef,
> Why do we need this ConductRDocRenderer?
> 
> —
> Reply to this email directly or view it on GitHub.
>